### PR TITLE
ARROW-12010: [C++][Compute] Improve performance of the hash table used in GroupIdentifier

### PR DESCRIPTION
**Pull request by michalursa**
This is the draft version of the code implementing functionality for mapping arbitrary set of input columns considered a key in grouping operation into a vector containing integer group identifiers (same combinations of input key columns get same ids). 

I will continue working on it and updating it with:
- integration with initial hash group by implementation in Arrow project, once it is finished and merged into master
- unit tests
- documentation

At this point group ids, row ids, offsets, hash values are 32-bit. The overflow checks are missing in current version and still need to be fixed. 

The entry point for id mapping is GroupBy class. It uses three main modules: storage defined in groupby_storage* files, hash defined in groupby_hash* files and hash table defined in groupby_map* files. Key values stored with the hash table are row oriented. Storage part of the code defines functions converting from column oriented storage to row oriented storage and back. It also implements comparison and appending keys to the incremental store.

I plan to add design doc in a form of a readme file later on.

The individual modules and functions present here have been tested with unit tests and are passing them but unit tests are not included in this change yet.

### CODE REVIEW

File: cpp/src/arrow/engine/util_avx2.cc [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+          _mm256_and_si256(indexes_8bit, _mm256_set1_epi8(0x0f)),
+          _mm256_and_si256(_mm256_srli_epi32(indexes_8bit, 4), _mm256_set1_epi8(0x0f)),
+          _mm256_set1_epi16(static_cast<uint16_t>(0xff00)));
+      __m256i input =
+          _mm256_loadu_si256(((const __m256i*)input_indexes) + 4 * i + loop_id);
+      // Shuffle bytes to get low bytes in the first 128-bit lane and high bytes in the
+      // second
+      input = _mm256_shuffle_epi8(
+          input, _mm256_setr_epi64x(0x0e0c0a0806040200ULL, 0x0f0d0b0907050301ULL,
+                                    0x0e0c0a0806040200ULL, 0x0f0d0b0907050301ULL));
```
Comment (bkietz) [inline cpp/src/arrow/engine/util_avx2.cc:111]: Please create `constexpr` variables for any magic numbers

File: cpp/src/arrow/engine/key_map.h [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+// ---------------------------------------------------
+// |     Empty bit*   |    Empty bit       |
+// ---------------------------------------------------
+// |   7-bit hash    |    7-bit hash      |
+// ---------------------------------------------------
+// * Empty bucket has value 0x80. Non-empty bucket has highest bit set to 0.
+// ** The order of bytes is reversed - highest byte represents 0th bucket.
+// No other part of data structure uses this reversed order.
+//
+class SwissTable {
```
Comment (bkietz) [inline cpp/src/arrow/engine/key_map.h:41]: Please add a higher level doc comment describing this class and it's utility. Specifically: describe how equality comparison and appending/storage of new entries are deferred to callbacks.

Separately, add a comment detailing its implementation and usage of stamps for vectorized probing.

File: cpp/src/arrow/engine/key_encode.h [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+    template <bool first_varbinary_col>
+    static void DecodeImp_avx2(uint32_t start_row, uint32_t num_rows,
+                               uint32_t varbinary_col_id, const KeyRowArray& rows,
+                               KeyColumnArray& col);
+#endif
+  };
+
+  class EncoderNulls {
+   public:
+    static void Encode(KeyRowArray& rows, const std::vector<KeyColumnArray>& cols,
```
Comment (bkietz) [inline cpp/src/arrow/engine/key_encode.h:425]: Please avoid mutable references. For out arguments, please use a mutable pointer
```suggestion
    static void Encode(KeyRowArray* rows, const std::vector<KeyColumnArray>& cols,
```

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/engine/key_encode.cc [hunk 4]
```diff
@@ -0,0 +1,1604 @@
...
+      case 1:
+        for (uint32_t i = 0; i < num_rows; ++i) {
+          row_base[i * row_size] = col_base[i];
+        }
+        break;
+      case 2:
+        for (uint32_t i = 0; i < num_rows; ++i) {
+          *reinterpret_cast<uint16_t*>(row_base + i * row_size) =
+              reinterpret_cast<const uint16_t*>(col_base)[i];
+        }
```
Comment (bkietz) [inline cpp/src/arrow/engine/key_encode.cc:453]: There's a lot of unaligned accesses like this one. This is undefined behavior in C++ and it's not supported on all platforms. Could we use SafeLoadAs and SafeStore? If those produce a performance regression, can we optimize them?

File: cpp/src/arrow/compute/exec/key_hash.cc [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+inline uint32_t Hashing::avalanche_helper(uint32_t acc) {
+  acc ^= (acc >> 15);
+  acc *= PRIME32_2;
+  acc ^= (acc >> 13);
+  acc *= PRIME32_3;
+  acc ^= (acc >> 16);
+  return acc;
+}
+
+void Hashing::avalanche(const arrow::internal::CpuInfo* cpu_info, uint32_t num_keys,
```
Comment (bkietz) [inline cpp/src/arrow/compute/exec/key_hash.cc:47]: If the intent is to allow functions to be called with any hardware flags independent of any other calls, it's worth noting that CpuInfo is a singleton. So for example disabling AVX2 for Hashing::avalanche will prevent it from being used in any other functions as well. I think what you want is to pass an `int64_t hardware_flags` which can be acquired with `CpuInfo::GetInstance()->hardware_flags()` or explicitly specified

**Review by bkietz — CHANGES_REQUESTED** [review]

**Comment by bkietz** [comment]
@michalursa looks like tests are hanging on our [bigendian CI](https://travis-ci.com/github/apache/arrow/jobs/506281680#L2907). Is this quick to address or should we leave this for follow up

**Comment by nealrichardson** [comment]
> @michalursa looks like tests are hanging on our [bigendian CI](https://travis-ci.com/github/apache/arrow/jobs/506281680#L2907). Is this quick to address or should we leave this for follow up

cc @kiszk

**Comment by michalursa** [comment]
> > @michalursa looks like tests are hanging on our [bigendian CI](https://travis-ci.com/github/apache/arrow/jobs/506281680#L2907). Is this quick to address or should we leave this for follow up
> 
> cc @kiszk

It requires a bit of thinking (or debugging) to make the code work with big endian. But since we still have the other group by implementation, for now I am disabling new one on big endian architectures, and will move this issue to a separate jira.

**Review by bkietz — APPROVED** [review]
LGTM, I'll merge.

Please add follow up JIRAs and link them here, including:
- Extract vectorized bit utilities
- Enable GrouperFastImpl for big endian platforms
- Unit tests should iterate through supported hardware flags (so an AVX2 build should test AVX2 and scalar implementations)

**Pull request closed without merging**